client/api/api.go

```diff
@@ -42,22 +43,25 @@ type errorReply struct {
 
 // CheckResponse returns an error (of type *Error) if the response
 // status code is not 2xx.
-func CheckResponse(res *http.Response) error {
-	if res.StatusCode >= 200 && res.StatusCode <= 299 {
+func CheckResponse(
+	res *http.Response,
+	ioReader IoReaderUtils,
```

Comment:
In general, tests should not change the signature of the function it tests.  You can simply replace the  line
```
ioutil.ReadAll(res.Body)
```
to be a function all
```
var readAll func(r io.Reader) ([]byte, error)
readAll(res.Body)
```

Reply:
Same for the other parameter

Reply:
You can also declare them as interfaces and mock the interface

Reply:
These dependencies are call from within this function. I think dependency injection style makes it easier to test and maintain. I agreed when we write test we shouldn't change the original function, but I don't think we can mock something that's called within the tested function do we?

Reply:
can you point me to the doc on what does "Go testing guideline" say about the Dependency injection style?  I don't normally see this type of "injection" in Go.

Reply:
The reason I ask is that Go is not an OO language so I have never seen any "dependency injection" in this PR.

Reply:
I think I saw that in the Go testing course Sean shared. The goal of that style was to isolate the dependencies. Your way would actually achieve the same purpose. I have never worked with Go but with web API developments this style is very popular and language isn't really the case (I saw this practice common in both NodeJS and Spring). Both those frameworks have built-in support for that though, so using a struct in Go was just my hack. But with that being said, both seems to achieve the same purpose so I am fine with switching style.

Reply:
The solution I showed in the first comment does not require the test to change the interface of the function it tests. The trick is to replace the function call from a direct call to a library to a call to a local function.  Please try it out and you will see how it works

Reply:
Synced with Ryan offline and this should be reflected in the PR now